esphome/components/shelly_dimmer/shelly_dimmer.cpp

```diff
@@ -0,0 +1,490 @@
+    std::memcpy(data + 4, payload, len);
+    pos += len;
+  }
+
+  // Calculate checksum for the payload.
+  const uint16_t csum = shelly_dimmer_checksum(data + 1, 3 + len);
+  data[pos++] = static_cast<uint8_t>(csum >> 8);
+  data[pos++] = static_cast<uint8_t>(csum & 0xff);
+  data[pos++] = SHELLY_DIMMER_PROTO_END_BYTE;
+  return pos;
+}
+
+int ShellyDimmer::handle_byte_(uint8_t c) {
+  const uint8_t pos = this->buffer_pos_;
+
+  if (pos == 0) {
+    // Must be start byte.
+    return c == SHELLY_DIMMER_PROTO_START_BYTE ? 1 : -1;
+  } else if (pos < 4) {
+    // Header.
+    return 1;
+  }
+
+  // Decode payload length from header.
+  const uint8_t payload_len = this->buffer_[3];
+  if ((4 + payload_len + 3) > SHELLY_DIMMER_BUFFER_SIZE) {
+    return -1;
+  }
+
+  if (pos < 4 + payload_len + 1) {
+    // Payload.
+    return 1;
+  }
+
+  if (pos == 4 + payload_len + 1) {
+    // Verify checksum.
+    const uint16_t csum = (this->buffer_[pos - 1] << 8 | c);
+    const uint16_t csum_verify = shelly_dimmer_checksum(&this->buffer_[1], 3 + payload_len);
+    if (csum != csum_verify) {
+      return -1;
+    }
+    return 1;
+  }
+
+  if (pos == 4 + payload_len + 2) {
+    // Must be end byte.
+    return c == SHELLY_DIMMER_PROTO_END_BYTE ? 0 : -1;
+  }
+  return -1;
+}
+
+bool ShellyDimmer::read_frame_() {
+  while (this->available()) {
+    const uint8_t c = this->read();
+    this->buffer_[this->buffer_pos_] = c;
+
+    ESP_LOGV(TAG, "Read byte: 0x%02x (pos %d)", c, this->buffer_pos_);
+
+    switch (this->handle_byte_(c)) {
+      case 0: {
+        // Frame successfully received.
+        this->handle_frame_();
+        this->buffer_pos_ = 0;
+        return true;
+      }
+      case -1: {
+        // Failure.
+        this->buffer_pos_ = 0;
+        break;
+      }
+      case 1: {
+        // Need more data.
+        this->buffer_pos_++;
+        break;
+      }
+    }
+  }
+  return false;
+}
+
+bool ShellyDimmer::handle_frame_() {
+  const uint8_t seq = this->buffer_[1];
+  const uint8_t cmd = this->buffer_[2];
+  const uint8_t payload_len = this->buffer_[3];
+
+  ESP_LOGD(TAG, "Got frame: 0x%02x", cmd);
+
+  // Compare with expected identifier as the frame is always a response to
+  // our previously sent command.
+  if (seq != this->seq_) {
+    return false;
+  }
+
+  const uint8_t *payload = &this->buffer_[4];
+
+  // Handle response.
+  switch (cmd) {
+    case SHELLY_DIMMER_PROTO_CMD_POLL: {
+      if (payload_len < 16) {
+        return false;
+      }
+
+      const uint8_t hw_version = payload[0];
+      // payload[1] is unused.
+      const uint16_t brightness = payload[3] << 8 | payload[2];
+
+      const uint32_t power_raw = payload[7] << 24 | payload[6] << 16 | payload[5] << 8 | payload[4];
+
+      const uint32_t voltage_raw = payload[11] << 24 | payload[10] << 16 | payload[9] << 8 | payload[8];
+
+      const uint32_t current_raw = payload[15] << 24 | payload[14] << 16 | payload[13] << 8 | payload[12];
```

Comment:
Just to reuse the helpers we have =)

```suggestion
      const uint32_t power_raw = encode_uint32(payload[7], payload[6], payload[5], payload[4]);

      const uint32_t voltage_raw = encode_uint32(payload[11], payload[10], payload[9], payload[8]);

      const uint32_t current_raw = encode_uint32(payload[15], payload[14], payload[13], payload[12]);
```